File: tests/module.cpp
#include "examples/module_base.hpp"
#include <catch2/catch.hpp>
#include <sde/detail_/module_pimpl.hpp>
#include <sde/module.hpp>

using namespace sde;

inline Module make_module() {
    auto ptr = std::make_shared<Rectangle>();
    return Module(std::make_unique<detail_::ModulePIMPL>(ptr));
}

inline Module make_prism() {
    auto ptr = std::make_shared<Prism>();
    return Module(std::make_unique<detail_::ModulePIMPL>(ptr));
}

TEST_CASE("Module : default ctor") {
    Module m;
    REQUIRE(!m.ready());
    REQUIRE(!m.locked());
    REQUIRE(m.inputs() == type::input_map{});
    REQUIRE(m.submods() == type::submodule_map{});
}

TEST_CASE("Module : Equality") {
    SECTION("Default state") {
        Module m1, m2;
        REQUIRE(m1 == m2);
        REQUIRE(!(m1 != m2));
    }
    SECTION("Non-default state") {
        Module m1 = make_module(), m2 = make_module();
        REQUIRE(m1 == m2);
        REQUIRE(!(m1 != m2));

        SECTION("Locked-ness") {
            m1.lock();
            REQUIRE(m1 != m2);
            REQUIRE(!(m1 == m2));
        }

        SECTION("Input values") {
            m1.change_input("Dimension 1", double{1.23});
            REQUIRE(m1 != m2);
            REQUIRE(!(m1 == m2));
        }
    }
}

TEST_CASE("Module : run_as") {
    Module m(make_module());
    auto[area] = m.run_as<Area>(1.23, 4.56);
    REQUIRE(area == 1.23 * 4.56);
}

TEST_CASE("Module : run") {
    Module m(make_module());
    auto inputs = m.inputs();
    inputs.at("Dimension 1").change(double{1.23});
    inputs.at("Dimension 2").change(double{4.56});
    auto rv = m.run(inputs);
    REQUIRE(rv.at("area").value<double>() == 1.23 * 4.56);
}

TEST_CASE("Module : ready") {
    Module m(make_module());
    REQUIRE(!m.ready());
    REQUIRE(m.ready<Area>());
}

TEST_CASE("Module : lock") {
    Module m(make_module());
    REQUIRE(!m.locked());
    m.lock();
    REQUIRE(m.locked());
}

TEST_CASE("Module : locked") {
    Module m(make_module());
    m.lock();
    REQUIRE(m.locked());
}

TEST_CASE("Module : change_input") {
    Module m(make_module());
    m.change_input("Dimension 1", 1.23);
    REQUIRE(m.inputs().at("Dimension 1").value<double>() == 1.23);
}

TEST_CASE("Module : change_submod") {
    Module m(make_prism());
    m.change_submod("area", std::make_shared<Module>(make_module()));
    REQUIRE(m.submods().at("area").value() == make_module());
}

TEST_CASE("Module : hash") {
    Module m(make_module());
    type::hasher h(bphash::HashType::Hash128);
    m.hash(h);
    auto hv = bphash::hash_to_string(h.finalize());
    REQUIRE(hv == "1b740793eacb5cdf38dc426f626ab5ab");
}

TEST_CASE("Module : copy ctor") {
    Module m1(make_module());
    Module m2(m1);
    REQUIRE(m1 == m2);
}

TEST_CASE("Module : copy assignment") {
    Module m1(make_module());
    Module m2;
    auto* ptr = &(m2 = m1);
    REQUIRE(m1 == m2);
    REQUIRE(ptr == &m2);
}

TEST_CASE("Module : move ctor") {
    Module m1(make_module());
    Module m2(m1);
    Module m3(std::move(m1));
    REQUIRE(m3 == m2);
}

TEST_CASE("Module : move assignment") {
    Module m1(make_module());
    Module m2(m1), m3;
    auto* ptr = &(m3 = std::move(m1));
    REQUIRE(m3 == m2);
    REQUIRE(ptr == &m3);
}

File: tests/module.py
from sde import *
from tests import *
import unittest

mm = sde.ModuleManager()
cppyy.gbl.load_modules(mm)

def get_rectangle():
    mod = mm.at("Rectangle")
    return sde.Module(mod)

def get_prism():
    mod = mm.at("Prism")
    return sde.Module(mod)

class test_module(unittest.TestCase):
    def setUp(self):
        self.r1 = get_rectangle()
        self.r2 = get_rectangle()
        self.p1 = get_prism()
        self.p2 = get_prism()

    def test_default_ctor(self):
        m = sde.Module()
        self.assertFalse(m.locked())

    def test_copy_ctor(self):
        m = sde.Module(self.r1)
        self.assertEqual(self.r1, m)

    def test_equality(self):
        m1 = sde.Module()
        m2 = sde.Module()
        self.assertEqual(m1, m2)

    def test_lockedness(self):
        self.assertEqual(self.r1, self.r2)
        self.r2.lock()
        self.assertNotEqual(self.r1, self.r2)

    def test_change_input(self):
        self.r1.change_input("Dimension 1", 1.23)
        self.assertNotEqual(self.r1, self.r2)

    def test_run_as(self):
        area = self.r1.run_as["Area"](1.23, 4.56)
        self.assertEqual(area[0], 1.23*4.56)

    def test_run(self):
        inputs = self.r1.inputs()
        inputs.at("Dimension 1").change(1.23)
        inputs.at("Dimension 2").change(4.56)
        rv = self.r1.run(inputs)
        self.assertEqual(rv.at("area").value["double"](), 1.23*4.56)

    def test_ready(self):
        self.assertFalse(self.r1.ready())
        self.assertTrue(self.r1.ready["Area"]())

    def test_lock(self):
        self.assertFalse(self.r1.locked())
        self.r1.lock()
        self.assertTrue(self.r1.locked())

    def test_change_submod(self):
        self.p1.change_submod("area", get_rectangle())
        self.assertEqual(self.p1.submods().at("area").value(), get_rectangle())

if __name__ == "__main__":
    unittest.main(verbosity=2)                      
